lib/bridge/QuitDebouncer.js

```diff
@@ -0,0 +1,125 @@
+/*eslint no-invalid-this: 0*/
+"use strict";
+var Promise = require("bluebird");
+
+function QuitDebouncer(ircBridge) {
+    this.ircBridge = ircBridge;
+
+    // Measure the probability of a net-split having just happened using QUIT frequency.
+    // This is to smooth incoming PART spam from IRC clients that suffer from a
+    // net-split (or other issues that lead to mass PART-ings)
+    this._debouncerForServer = {
+      // $server.domain: {
+            // $nick: {
+            //   promise: Promise, //Promise that resolves if the user joins a channel having quit
+            //   resolve: Function //function to call when a user joins
+            // }
+      // }
+    };
+
+    // Keep a track of the times at which debounceQuit was called, and use this to
+    // determine the rate at which quits are being received. This can then be used
+    // to detect net splits.
+    /*$server : [1477386173850, 1477386173825, ...]*/
+    this._quitTimesMs = {};
+    Object.keys(this.ircBridge.config.ircService.servers).forEach((domain) => {
+        this._quitTimesMs[domain] = [];
+    });
+}
+
+/**
+ * Called when the IrcHandler receives a JOIN. This resolves any promises to join that were made
+ * when a quit was debounced during a split.
+ * @param {string} nick The nick of the IRC user joining.
+ * @param {IrcServer} server The sending IRC server.
+ */
+QuitDebouncer.prototype.onJoin = function (nick, server) {
+    if (!this._debouncerForServer[server.domain]) {
+        return;
+    }
+    let rejoin = this._debouncerForServer[server.domain][nick];
+    if (rejoin) {
+        rejoin.resolve()
+    }
+}
+
+/**
+ * Debounce a QUIT received by the IrcHandler to prevent net-splits from spamming leave events
+ * into a room when incremental membership syncing is enabled.
+ * @param {Request} req The metadata request.
+ * @param {IrcServer} server The sending IRC server.
+ * @param {string} matrixUser The virtual user of the user that sent QUIT.
```

Comment:
This is done (also it would be helpful if you could click the [+] on a line that will definitely change if the comment requires it (also, I prefer the new style of review - it supports 'outdated' now))